Use OpenOCD for Launchxl build and sanitize

alevy opened the pull request:
### Pull Request Overview

This pull request makes a number of changes to the `launchxl` board's build system:

- Replaces TI's proprietary tool (that's a nightmare to setup) with OpenOCD (which is slightly less of a nightmare) to flash the kernel over USB by default.

- Moves the process ROM offset to that expected by `tockloader` by default. There is still plenty of room for the kernel, and doing this makes programming apps using `tockloader` much simpler.

- Removes the `CCFG` registers from the kernel binary, and instead uses a _separate_ binary for encoding and flashing the `CCFG` registers as a separate step. This one's important. Overwriting the CCFG registers is both dangerous (it can brick the board) and very very instance specific (i.e. this is where you'd put the MAC address). Having it as a separate step allows the user to just do it once when they buy the board from TI and spend most of their time not touching it. It also gives a more controlled environment when changing the registers later (e.g. for deployment, or to fiddle with power settings).

- Adds an optional `flash-jlink` and `flash-ccfg-jlink` make targets for programming with JLinkExe and an external debugger as well.

I also updated the launchxl README to reflect flashing with OpenOCD and JLinkExe and how to specifically set up the jumpers for each.

One nit to note is that OpenOCD doesn't properly support hardware reset for the on-board debug chip yet, so we're using soft-reset instead---this _only_ resets the Cortex-M core, and could leave the rest of the hardware in an unexpected state. Always best to manually tap the reset button next to the USB port after flashing the kernel.

### Testing Strategy

Fairly exhaustive testing on both the CC2652 launchxl and CC1312 launchxl, including bricking two of my four boards when I was blindly using the previous kernel binary.

### TODO or Help Wanted

(Specifically for @ppannuto) I changed Makefile.common to be more permissive about which binaries it's willing to convert to other formats. I think the changes are pretty straight forward and safe, but input would be useful none-the-less.

This PR also removes the Makefile.common targets for converting to `.hex` files since `llvm-objcopy` _only_ supports converting to binary. It however doesn't cleanup references to those targets from various custom scripts and Makefile targets in a couple boards.

To come in future a PRs:

- ~~Flashing using JLinkExe using an external debugger (JLinkExe doesn't support the on-board debugger, and probably never will)~~ _Done_

- Documenting the CCFG registers set in `ccfg.rs`. This is _not_ gonna be fun, and doing something sane seems more useful short term then waiting to build up the motivation to try and understand that section of the data-sheet properly.

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make formatall`.

alevy commented:
/cc @JayKickliter @refugeesus

bradjc commented on boards/Makefile.common:
```diff
@@ -164,3 +165,6 @@ clean::
 debug: target/$(TARGET)/debug/$(PLATFORM).elf
+
+.PHONY: release
+debug: target/$(TARGET)/release/$(PLATFORM).elf
```
Presumably this should be `release:`

Also, why do we have these targets in two places (also line 108)? That seems odd.

alevy replied:
You're right. fixing.

ppannuto commented on boards/Makefile.common:
```diff
@@ -124,34 +127,35 @@ target:
 	$(Q)$(OBJCOPY) -O=binary $^ $@
 
-target/$(TARGET)/debug/$(PLATFORM).bin: target/$(TARGET)/debug/$(PLATFORM).elf
+target/$(TARGET)/debug/%.bin: target/$(TARGET)/debug/%.elf
```
If we're going to go for pattern rules, why not just go all the way and replace both of these with
`%.bin: %.elf`

alevy replied:
Right, you're the makefile expert :) I'm entirely sure what might possibly break if I change them that drastically, but if you think it's ok, I'm game.

ppannuto commented on boards/Makefile.common:
```diff
@@ -124,34 +127,35 @@ target:
-target/$(TARGET)/debug/$(PLATFORM).lst: target/$(TARGET)/debug/$(PLATFORM).elf
-	$(Q)$(OBJDUMP) $(OBJDUMP_FLAGS) $< > target/$(TARGET)/debug/$(PLATFORM).lst
+target/$(TARGET)/debug/%.lst: target/$(TARGET)/debug/%.elf
+	$(Q)$(OBJDUMP) $(OBJDUMP_FLAGS) $< > $@
```
Same here for `.lst` -> `.elf`

ppannuto commented on boards/Makefile.common:
```diff
@@ -124,34 +127,35 @@ target:
 
 target/$(TARGET)/debug/$(PLATFORM).elf: target/$(TARGET)/debug/$(PLATFORM)
-	$(Q)cp target/$(TARGET)/debug/$(PLATFORM) target/$(TARGET)/debug/$(PLATFORM).elf
+	$(Q)cp $< $@
```
And for `%.elf` -> `%`

bradjc commented:
Is there value to having the ccfg.rs file in the chip crate still?

alevy commented:
> Is there value to having the ccfg.rs file in the chip crate still?

I think it will be valuable to have a `ccfg.rs` file that actually has register definitions (as opposed to just an array of constants), but _not_ an instantiation.

refugeesus commented:
I think implementing a minimal ccfg with its register definitions is probably smart since the ccfg file is supposed to make it easier for users to make broad changes to radio settings, clock, and power controls. I am not 100% certain what a setting in ccfg pertaining to `vd_ctl` does when I try to change something about `vd_ctl` independently (perhaps blocking registers?).

bradjc commented:
Should this PR then delete the ccfg.rs file that currently exists?

refugeesus commented:
No, you must have some ccfg settings or the board will brick. The array of constants forces it to work but the settings should be implemented later. Maybe not in this PR necessarily though.

bradjc commented:
Why is 23ac2dd01d78a93ca83f322f1dd72da0b18d2e91 attached to this PR?

And I guess what I am asking is how is https://github.com/tock/tock/blob/master/chips/cc26xx/src/ccfg.rs being used after this change? It seems like the CCFG struct is defined with the board now.

alevy commented:
@bradjc it's not exactly being used by the board. The board has an _additional_ binary target (see cargo.toml) that constains _just_ the ccfg struct and is programmed _separately_ only when you first use the board (`make flash-ccfg`)

alevy commented:
After an in depth conversation with @bradjc we discovered that good condition was because I hadn't deleted ccfg.rs from the chip crate, even though I removed it from lib.rs and said I removed it.

Will fix

refugeesus commented:
One thing I noticed flashing different boards was that due to a reset bug in openocd, a board that has been flashed with a different AON configuration than the TI distribution will not be able to reset itself after being switched into DFU mode per TI's firmware update instructions [here](http://processors.wiki.ti.com/index.php/XDS110#Updating_the_XDS110_Firmware). However you can still update the firmware by building one of the base RTOS distributions from TI in CCS and flash the board which will prompt CCS to run the firmware update itself which works.

alevy commented:
@bradjc fixed `ccfg.rs` issue

bradjc commented on boards/launchxl/src/ccfg.rs:
```diff
@@ -0,0 +1,23 @@
+#![no_std]
+#![no_main]
+#![feature(used, panic_implementation)]
+#![feature(used, panic_implementation)]
```
Delete

alevy replied:
done

ppannuto commented:
Re makefile: Yeah, sorry, once I started cleaning that up I kept going, and didn't want to create a somewhat needless conflict (at first), which is a hole that dug deeper as I did more.
